### ISSUE TITLE
(IV_MIN << 0) returns zero under "use integer"

### ISSUE THREAD

**Issue by t-a-k** [Perl/perl5#18639]
**Description**
`$x << 0` results in `0` if `use integer` is in force and `$x` is equal to `IV_MIN`.

**Steps to Reproduce**
```
$ perl -wle 'my $x = -(~0 >> 1) - 1; print $x; { use integer; print $x << 0 }'
-2147483648
0
```
`$x` shall be equal to `IV_MIN` here.

**Expected behavior**
`$x << 0` should return `$x` for any IVs.

**Perl configuration**
```
Summary of my perl5 (revision 5 version 33 subversion 8) configuration:
  Commit id: cb174a785cca1bace5d049d08be208784028e671
  Platform:
    osname=linux
    osvers=4.19.0-14-686-pae
    archname=i686-linux
    uname='linux x41 4.19.0-14-686-pae #1 smp debian 4.19.171-2 (2021-01-30) i686 gnulinux '
    config_args='-de -Dusedevel -DEBUGGING'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=undef
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=4
    doublesize=8
    byteorder=1234
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=12
    longdblkind=3
    ivtype='long'
    ivsize=4
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=4
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/i686-linux-gnu/8/include-fixed /usr/include/i386-linux-gnu /usr/lib /lib/i386-linux-gnu /lib/../lib /usr/lib/i386-linux-gnu /usr/lib/../lib /lib /lib64 /usr/lib64
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.28.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.28'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    DEBUGGING
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under linux
  Compiled at Mar 10 2021 23:55:51
  @INC:
    /usr/local/lib/perl5/site_perl/5.33.8/i686-linux
    /usr/local/lib/perl5/site_perl/5.33.8
    /usr/local/lib/perl5/5.33.8/i686-linux
    /usr/local/lib/perl5/5.33.8
```

**Comment by sisyphus** [comment]
I think that the C89 spec for  the<code><<</code> operator, supports your expectation.

However, by my reading, the C99 and C11 specs imply that *every* left-shift operation of a signed, negative value is undefined behaviour.
AFAICS, the wording does not provide any allowance for the special case where a 0 bits left shift  of a negative value is being performed.

I don't know how that should impact on perl. (I'm never sure what the rules are.)

It does, however, strike me as being intuitively obvious that a left-shift operation of 0 bits should result in an unchanged value - just as you expect.

The C99 and C11 spec:
```
The result of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are filled with zeros.
If E1 has an unsigned type, the value of the result is E1 * (2 ** E2), reduced modulo
one more than the maximum value representable in the result type.
If E1 has a signed type and nonnegative value, and E1 * (2 ** E2) is representable in the result
type, then that is the resulting value; otherwise, the behavior is undefined.
```

**Comment by tonycoz** [comment]
Looks like this was introduced in 814735a391b, which added:

```
+    if (left) {
+        if (iv == IV_MIN) { /* Casting this to a UV is undefined behavior */
+            return 0;
+        }
+        return (IV) (((UV) iv) << shift);
+    }
```

but this comment seems incorrect to me.  @khw?

**Comment by khwilliamson** [comment]
I was conflating in my mind casting and complementing.  You can't complement IV_MIN on a 2's complement machine.

I believe PR https://github.com/Perl/perl5/pull/18644 fixes the problem

**Comment by tonycoz** [comment]
Fixed by 4e82c85b1c9c9b30253b8624470da6f20a6c0604
